### DISCUSSION TITLE
Handle the stripe subscriptions process with filament

### DISCUSSION THREAD

**Discussion by devhoussam1998** [filamentphp/filament#12612] [Help]
### Package

Actions

### Package Version

v3.2.71

### How can we help you?

I'm using laravel cashier with stripe I setup everything correctly but I face some issues.

I get this error message when I click on orderNow action

"The payment attempt failed because of an invalid payment method."

Here is what I did:

**ListProducts.php**

```
<?php
 
namespace App\Livewire;
 
use App\Models\Shop\Product;
use Filament\Forms\Concerns\InteractsWithForms;
use Filament\Forms\Contracts\HasForms;
use Filament\Tables\Columns\TextColumn;
use Filament\Tables\Concerns\InteractsWithTable;
use Filament\Tables\Contracts\HasTable;
use Filament\Tables\Table;
use Illuminate\Contracts\View\View;
use Livewire\Component;
 
class ListProducts extends Component implements HasForms, HasTable
{
    use InteractsWithTable;
    use InteractsWithForms;
    
    public function table(Table $table): Table
    {
        return $table
            ->query(Product::query())
            ->columns([
                TextColumn::make('name'),
            ])
            ->actions([
                Action::make('orderNow')
                ->action(function (Product $record) {
                    $this->subscribeToPlan($record->product_id);
                }),
            ]);
    }

    public function subscribeToPlan($productId)
    {
        $user = auth()->user();

        $user->newSubscription('default', $productId)->create();
    }
    
    public function render(): View
    {
        return view('livewire.list-products');
    }
}
```

**ProductSeeder.php**

```
<?php

namespace Database\Seeders;

use App\Models\Product;
use Illuminate\Database\Seeder;

class ProductSeeder extends Seeder
{
    /**
     * Run the database seeds.
     */
    public function run(): void
    {
        $products = [
            [
                'name' => 'Basic',
                'product_id' => 'price_xxxxxxxxxxxxxxxxxxxxx',
            ],
            [
                'name' => 'Advanced',
                'product_id' => 'price_xxxxxxxxxxxxxxxxxxxxx',
            ],
        ];

        foreach ($products as $product) {
            Product::create($product);
        }
    }
}
```

**Comment by hzeeshan** [comment]
Late to this but in case anyone finds it via search: that "invalid payment method" error is because `->create()` is being called with no payment method attached.

Cashier can't charge a subscription without a real Stripe `PaymentMethod`. Clicking a button that just runs `newSubscription()->create()` won't work on its own, you need to collect card details on the frontend first (Stripe Elements), get back a `pm_...` token, and pass that in:

```php
$user->newSubscription('default', $priceId)->create($paymentMethodId);
```

If you just want to confirm the backend works before building the card form, `->create('pm_card_visa')` with a test token does the trick.

One more thing I'd double check: you're passing `$record->product_id` into `newSubscription()`. Cashier wants the **price** ID there (`price_...`), not the product ID (`prod_...`). I've been bitten by that exact mixup before.

Btw if you get past this and end up needing to actually manage those subscriptions later (cancel, resume, swap, see MRR) I got tired of rebuilding that admin side every project so I packaged it up: https://hafiz.dev/blog/filament-billing-the-missing-stripe-admin-layer-for-your-panel
